[FEATURE] [API] Créer une variable d'environnement pour définir les claims OIDC supplémentaires à sauvegarder (PIX-9379)

lego-technix opened the pull request:
## :unicorn: Problème

Pour certains partenaires SSO, il y a le nouveau besoin de stocker certaines informations (_claims_ OIDC) sur l'utilisateur, lors de la connexion de l'utilisateur par le SSO. Et ces informations sont différentes suivant chaque partenaire SSO.

## :robot: Proposition

Ajouter une option de configuration `claimsToStore` remplie à partir de la variable d’environnement `${IDP}_CLAIMS_TO_STORE` (à définir pour chaque SSO OIDC) pour définir les _claims_ à sauvegarder.

Les _claims_ définis par `${IDP}_CLAIMS_TO_STORE` sont des _claims_ **obligatoires**. Si un de ces _claims_ n’est pas récupéré à partir des _userInfo_, alors l’authentification échoue.

Exemple : 

```shell
FWB_CLAIMS_TO_STORE=employeeNumber,studentGroup
```

Les _claims_ spécifiés ne doivent pas obligatoirement être des _claims_ OIDC standards, même si c’est mieux. Pour info voici la liste des _claims_ OIDC standards : 
https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims

On précise bien qu’un compte utilisateur ne pourra avoir qu’une seule valeur pour un _claim_ donné. Par exemple un compte utilisateur ne pourra avoir qu’un `employeeNumber`.

## :rainbow: Remarques

### Refactors de préparation

Les premiers commits de cette PR sont des petits correctifs et des petits réarrangements pour préparer la place aux changements plus importants, qui vont se charger du stockage des informations utilisateurs supplémentaires, et qui vont avoir lieu dans une PR suivante.

### Facilitation de la gestion de tous les cas pour claimsToStore

Enfin, une remarque sur le code suivant qui a été mis dans le constructeur de `OidcAuthenticationService` : 
```javascript
if (!lodash.isEmpty(claimsToStore)) {
  this.claimsToStore = claimsToStore;
}
```

Ce code permettra dans le code à venir de faire un simple test `if (this.claimsToStore)` chaque fois qu'on voudra savoir s'il faut gérer le cas où il y a des `claimsToStore`, sans avoir à tester tous les cas possibles à chaque fois. Car un tableau vide est aussi un [Truthy](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) en JavaScript.

## :100: Pour tester

Au niveau des RA il n'est pas facile de tester les SSO, donc on ne teste pas

En local : 

1. Définir la variable d'environnement suivante : 

   ```shell
   FWB_CLAIMS_TO_STORE=employeeNumber,studentGroup
   ```
2. L'ajout de cette variable n'entraine, au stade de cette PR, aucun changement de comportement. Donc tester uniquement tous les SSO et vérifier qu'il n'y a pas de régression

lego-technix committed:
feat(api): add true getArrayOfStrings and rename

previous getArrayOfStrings into getArrayOfUpperStrings

lego-technix committed:
feat(api): add new claimsToStore property to

OidcAuthenticationService

lego-technix committed:
fix(api): remove useless await for jsonwebtoken.decode

lego-technix committed:
refactor(api): remove useless userInfoUrl param passing

lego-technix committed:
test(api): remove useless await

getUserInfoMissingFields is synchronous

lego-technix committed:
feat(api): leverage new claimsToStore prop and getArrayOfStrings function for FWB

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr7387.review.pix.fr)
  * [App (.org)](https://app-pr7387.review.pix.org)
  * [Orga (.fr)](https://orga-pr7387.review.pix.fr)
  * [Orga (.org)](https://orga-pr7387.review.pix.org)
  * [Certif (.fr)](https://certif-pr7387.review.pix.fr)
  * [Certif (.org)](https://certif-pr7387.review.pix.org)
  * [1D](https://1d-pr7387.review.pix.fr)
  * [Admin](https://admin-pr7387.review.pix.fr)
  * [API](https://api-pr7387.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr7387.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr7387/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr7387/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr7387/environment)

igorissen commented on api/lib/domain/services/authentication/oidc-authentication-service.js:
```diff
@@ -191,7 +191,62 @@ class OidcAuthenticationService {
+    return redirectTarget.toString();
+  }
+
+  async _getUserInfoFromEndpoint({ accessToken }) {
```
Est-ce que ce serait possible de séparer en 2 ?
- la déplacement de toutes les fonctions privées en bas
- passage d'une fonction public en privée

lego-technix replied:
Je vais séparer en 2 commits.

lego-technix replied:
Voilà la modification a été séparée en 2 commits. C'est vrai que cela rend la lecture des modifications effectuées plus claires 👍

igorissen commented on api/tests/unit/domain/services/authentication/oidc-authentication-service_test.js:
```diff
@@ -203,13 +203,13 @@ describe('Unit | Domain | Services | oidc-authentication-service', function () {
   });
 
-  describe('#getUserInfoMissingFields', function () {
+  describe('#_getUserInfoMissingFields', function () {
```
Est-ce que l'on teste les méthodes privées ? 🤔

lego-technix replied:
Attention, comme l'explique le commentaire du commit en question, c'est uniquement du déplacement de fonctions pour mieux comprendre l'intention du code. Ce n'est pas l'utilisation de la fonction native private de JavaScript.

lego-technix replied:
Et effectivement on ne pourrait pas tester une _vraie_ méthode privée 😃

igorissen reviewed (approved):
![](https://media.giphy.com/media/13zeE9qQNC5IKk/giphy.gif)

lego-technix committed:
refactor(api): rename "internal" functions

(i.e. not used from outside) with a "_" to identify them for clarity.

It's not the native JavaScript private feature because those functions are at
the moment tested outside of the class. So it's not a total refactoring, this
commit is for clarity.

lego-technix committed:
refactor(api): put "internal" functions at the end of module

This move is not a total refactoring, this is done for clarity.

EmmanuelleBonnemay reviewed (approved):
Testé fonctionnellement avec succès sur Firefox.
